### ISSUE TITLE
Not updating?

### ISSUE THREAD

**Issue by lysin** [mileswd/mac2imgur#41]
I did a fresh install of yosemite and downloaded mac2imgur, but I'm still at version 115. None of your latest commits appear to be there, and I see no option to force an update.

Any suggestions?

**Comment by BT23** [comment]
The latest release on here is from October. You have to download the project and compile it in XCode to get the latest version. I just compiled it a few minutes ago. Download it [here](http://www.filedropper.com/mac2imgurapp).

**Comment by mileswd** [comment]
I was intending to push an update a while, but Apple have released Swift 1.2, so I am updating the project and fixing a few things before I push out the new update. Trust me, it'll be worth the wait! :)

Update ETA is over the next few days.

**Comment by scalp42** [comment]
Any chance someone could build it please ? @BT23 link doesn't appear to be working for me.

**Comment by BT23** [comment]
@scalp42 [Here you go](http://www.filedropper.com/mac2imgurapp).

**Comment by scalp42** [comment]
Thanks a lot @BT23 but unfortunately it's not uploading the screenshots automatically for me.

**Comment by PunKeel** [comment]
@scalp42 Sorry to ask, but do you have XCode installed ? (Not telling you to install it, just asking)

Thanks!

@mileswd It looks like the events are not fired with the `NSPredicate` `kMDItemIsScreenCapture = 1` for people without XCode installed.
I'll look further into it …

**Comment by pcamp123** [comment]
It's not auto-uploading the screenshots for me either, I have XCode installed. I'm on 10.10.2, but thinking of going to 10.10.3. I can't get any screenshot utility to work, not Dropbox's upload, or mac2imgur, or anything.
Starting to think Apple broke something with 10.10.2...

**Comment by PunKeel** [comment]
fyi, Dropbox screenshot's feature worked on the other computer, but installing XCode does not seem to fix the issue.

On some IRC channel, they told me the monitoring object could be zombified. I haven't found exactly how to check it, but I think it's not the issue.

I've even rewrote the monitor thing in a closure, without success (works on my computer, not on the other one)
```swift
        query = NSMetadataQuery()
        
        // Only accept screenshots
        query.predicate = NSPredicate(format: "kMDItemIsScreenCapture = 1", argumentArray: nil)
        
        // Add observers
        NSNotificationCenter.defaultCenter().addObserverForName(NSMetadataQueryDidUpdateNotification, object: query, queue: NSOperationQueue.mainQueue()) { notification in
            NSLog("Event")
            if let itemsAdded = notification.userInfo?["kMDQueryUpdateAddedItems"] as? [NSMetadataItem] {
                for item in itemsAdded {
                    // Get the path to the screenshot
                    if let screenshotPath = item.valueForAttribute(NSMetadataItemPathKey) as? String {
                        let screenshotName = screenshotPath.lastPathComponent.stringByDeletingPathExtension
                        
                        // Ensure that the screenshot detected is from the right folder and isn't blacklisted
                        if screenshotPath.stringByDeletingLastPathComponent.stringByStandardizingPath == self.getScreenshotDirectory().stringByStandardizingPath && !contains(self.blacklist, screenshotName) {
                            println("Screenshot file event detected @ \(screenshotPath)")
                            delegate.screenshotDetected(screenshotPath)
                            self.blacklist.append(screenshotName)
                        }
                    }
                }
            }
        }
        // Start query
        query.startQuery()
```

**Comment by pcamp123** [comment]
What IRC channel? If it's #MacOSX on Freenode, I'm in there right now.. As well as a number of other channels, and would gladly help debug it any way I can..

**Comment by PunKeel** [comment]
It was on Freenode, channels #mac-dev & #swift-lang (but got no answer on the latter)

**Comment by mileswd** [comment]
@PunKeel The ```ScreenshotMonitor``` is retained as an object in ```AppDelegate``` and wouldn't be released until application close, I think.

Anyways, I've struggled with replicating this problem consistently - simply put, sometimes screenshots are detected, yet sometimes not.

Although I am not entirely sure, I believe this is actually an OS X problem, which I think is most likely related to some sort of change Apple have made to ```mds``` (the system metadata server) which is used for indexing files (how Spotlight returns results so quickly!)

To test if this really is the case, try this:
1. Kill the ```mds``` process (or any other ```Metadata.framework``` related processes)
2. Restart the ```ScreenshotMonitor``` (or mac2imgur)

Using a ```NSMetadataQuery``` to detect screenshots has always been a bit of a pain to be honest, although it's a simple and efficient method, dodgy things always seem to happen... If I can't find a clear-cut solution to this and a few other issues, I might just re-write the monitor to use ```FSEvents``` instead.

**Comment by mileswd** [comment]
@PunKeel By the way, I'd just like to say thanks for the help so far - much appreciated! :)

**Comment by PunKeel** [comment]
Thanks for the reply
> The ScreenshotMonitor is retained as an object in AppDelegate and wouldn't be released until application close, I think.
I know, but prefer to test stupid things instead of doing nothing :(

> thanks for the help so far
You're welcome (I'm sorry I didn't PR all my changed, I've done some pretty useful things ... [drag&drop, a progress bar ... I'll see to merge them])

> To test if this really is the case, try this:
Said to my beta tester, I'll come back with the results :dancer: 

As for the NSMetadataQuery, I'm not sure about how it works. I've seen (and think it's a good idea, was going to do it if I didn't see your repository) people watching the `Screenshots` folder for new files, but according to what you say, it would be the same thing
(and would not allow to change the folder while mac2imgur is running)

meh. I might also learn to read what you write up to the last line ...

**Comment by PunKeel** [comment]
Oh, nice spot @mileswd. 
Killing mds and restarting the app makes it work :+1:

**Comment by mileswd** [comment]
@PunKeel I'm glad that we've been able to pinpoint the issue to something :)

I never explained my choice in using a NSMetadataQuery, but you are exactly right: my reasoning was that it would allow people to change the location screenshots are saved to (and also allow interaction with other custom capturing applications) without having to restart or configure the application. Also at the time, ```FSEvents``` was unaccessible from Swift and therefore would have required an Objective-C library (however I think this is no longer the case).

I've also noticed a few ```mds``` crashes and high CPU usage by the process in this beta, so I assume Apple are currently working on this. Hopefully, there will be some improvements in an upcoming release :)

**Comment by PunKeel** [comment]
For information, to anyone not able to upload, you may open the terminal and run this:
`sudo launchctl list com.apple.metadata.mds PID`
It will return something like `"PID" = 72446;``
Then simply run `sudo kill -9 72446`, it will promp you for your password. And then the mds service will restart, and your screenshots will be uploaded

**Comment by mileswd** [comment]
I believe ```sudo killall mds``` may be a slightly simpler way of doing this :)

**Comment by PunKeel** [comment]
Just a little simpler, then. :-)

**Comment by pcamp123** [comment]
I've tried all manners of killing MDS, by using killall or by using the PID. I've done both with mac2imgur open or closed, and I've rebooted mac2imgur after doing it when it was open. Still not auto-uploading for me, on 10.10.3.

**Comment by scalp42** [comment]
Not working either on 10.9.5 :cry:

**Comment by mileswd** [comment]
Going to close this issue, as the original request of an update was satisfied 2 weeks ago.

@pcamp123 @scalp42 @PunKeel I no longer experience screenshots not being detected, I think Apple patched the issue in one of the 10.10.3 betas. However if any of you are still experiencing this, there is a separate issue (#46) where we can try and figure out what's happening.
